test/tester.go

```diff
@@ -0,0 +1,206 @@
+	db, err := database.New(ctx, DBTestCredentials(), logger)
+	if err != nil {
+		cancel()
+		return nil, errors.AddContext(err, "failed to connect to the DB")
+	}
+
+	// Start a noop mail sender in a background thread.
+	sender, err := email.NewSender(ctx, db, logger, &DependencySkipSendingEmails{}, FauxEmailURI)
+	if err != nil {
+		cancel()
+		return nil, errors.AddContext(err, "failed to create an email sender")
+	}
+	sender.Start()
+
+	// The meta fetcher will fetch metadata for all skylinks. This is needed, so
+	// we can determine their size.
+	mf := metafetcher.New(ctx, db, logger)
+
+	// The server API encapsulates all the modules together.
+	server, err := api.New(db, mf, logger, email.NewMailer(db))
+	if err != nil {
+		cancel()
+		return nil, errors.AddContext(err, "failed to build the API")
+	}
+
+	// Start the HTTP server in a goroutine and gracefully stop it once the
+	// cancel function is called and the context is closed.
+	srv := &http.Server{
+		Addr:    ":" + testPortalPort,
+		Handler: server.Router(),
+	}
+	go func() {
+		_ = srv.ListenAndServe()
+	}()
+	go func() {
+		select {
+		case <-ctx.Done():
+			_ = srv.Shutdown(context.TODO())
+		}
+	}()
+
+	return &AccountsTester{
+		Ctx:    ctx,
+		DB:     db,
+		Logger: logger,
+		cancel: cancel,
+	}, nil
+}
+
+// Get executes a GET request against the test service.
+func (at *AccountsTester) Get(endpoint string, params map[string]string) (r *http.Response, body []byte, err error) {
+	return at.executeRequest(http.MethodGet, endpoint, params, nil)
+}
+
+// Delete executes a DELETE request against the test service.
+func (at *AccountsTester) Delete(endpoint string, params map[string]string) (r *http.Response, body []byte, err error) {
+	return at.executeRequest(http.MethodDelete, endpoint, params, nil)
+}
+
+// Post executes a POST request against the test service.
+func (at *AccountsTester) Post(endpoint string, params map[string]string, postParams map[string]string) (r *http.Response, body []byte, err error) {
+	vals := url.Values{}
+	for k, v := range postParams {
+		vals[k] = append(vals[k], v)
+	}
+	serviceURL := testPortalAddr + ":" + testPortalPort + endpoint + "?" + encodeValues(params)
+	req, err := http.NewRequest(http.MethodPost, serviceURL, strings.NewReader(vals.Encode()))
+	if err != nil {
+		return nil, nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	if at.Cookie != nil {
+		req.Header.Set("Cookie", at.Cookie.String())
+	}
+	c := http.Client{}
+	r, err = c.Do(req)
+	if err != nil {
+		return
+	}
+	return processResponse(r)
+}
+
+// Put executes a PUT request against the test service.
+func (at *AccountsTester) Put(endpoint string, params map[string]string, putParams map[string]string) (r *http.Response, body []byte, err error) {
+	return at.executeRequest(http.MethodPut, endpoint, params, putParams)
+}
+
+// Shutdown performs a graceful shutdown of the AccountsTester service.
+func (at *AccountsTester) Shutdown() {
```

Comment:
This is not blocking for anything to actually shut down so I'd call it `Close() error` to implement the `io.Closer` interface and make people aware that this needs to be closed. It's easier to oversee a `Shutdown` method and it's also very unique to things like the `http.Server` which need to do some more work for a shutdown.

Reply:
I don't see how anyone will notice that this is an `io.Closer` without looking through the methods. And if they look through the methods and don't call the `Shutdown` one then I'm not sure a `Close` one would help.

But sure. It wouldn't hurt to try.